### ISSUE TITLE
tf.keras.backend.set_learning_phase doesn't work during evaluating model

### ISSUE THREAD

**Issue by CasiaFan** [tensorflow/tensorflow#15367]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: pip install tensorflow-gpu
- **TensorFlow version (use command below)**: 1.4.0
- **Python version**: 3.5.2
- **Exact command to reproduce**:

### Describe the problem
When I run following script，error  `InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'block1_conv1_bn/keras_learning_phase' with dtype bool`  occured.  After searching in keras repository and stackoverflow, I find it is caused by the design of learning phase parameter of BN layers which behave differently at training and testing time (See [here](https://blog.keras.io/keras-as-a-simplified-interface-to-tensorflow-tutorial.html)). fchollet have added `K.set_learning_phase()` for tensorflow to solve this problem. So when I use `keras` instead of `tf.keras`, no issue is reported. I wonder if this part is still not integrated into tensorflow completely. 

### Source code / logs
```python
import tensorflow as tf
# import keras  # work normally
keras = tf.keras
Xception = keras.applications.Xception
from keras import backend as K

data_path = "val.tfrecord"
if not isinstance(data_path, (tuple, list)):
    data_path = [data_path]
feature = {'image/encoded': tf.FixedLenFeature([], tf.string),
           'image/class_id': tf.FixedLenFeature([], tf.int64)}
# create file queue
filename_queue = tf.train.string_input_producer(data_path)
# tfrecord file reader
reader = tf.TFRecordReader()
_, example_string = reader.read(filename_queue)
# decode record
features = tf.parse_single_example(example_string, features=feature)
image = tf.decode_raw(features['image/encoded'], out_type=tf.uint8)
image = tf.cast(image, dtype=tf.float32)
# restore shape
image = tf.reshape(image, (96, 96, 3))
label = tf.cast(features['image/class_id'], dtype=tf.int64)
image_batch, label_batch = tf.train.shuffle_batch(tensors=[image, label],
                                                  batch_size=128,
                                                  capacity=10000,
                                                  min_after_dequeue=3000,
                                                  num_threads=8,
                                                  allow_smaller_final_batch=True)

# convert label to one hot label
sess = K.get_session()
# declare learning phase for BN/dropout
K.set_learning_phase(0)
label_batch = tf.one_hot(label_batch, 2, dtype=tf.float32)
model_input = keras.layers.Input(tensor=image_batch)
base_model = Xception(include_top=True,
                      weights=None,   # no pre-trained weights used
                      pooling="avg",
                      input_shape=(96, 96, 3),  # modify first layer
                      classes=2)
model_output = base_model(model_input)
test_model = keras.models.Model(inputs=model_input, outputs=model_output)
test_model.load_weights("weights.h5")
optimizer = tf.train.RMSPropOptimizer(learning_rate=2e-3, decay=0.9)
test_model.compile(optimizer=optimizer,
                   loss="categorical_crossentropy",
                   metrics=['accuracy'])
acc_value = keras.metrics.categorical_accuracy(label_batch, model_output)
# Fit model using data from tf records queue
coord = tf.train.Coordinator()
threads = tf.train.start_queue_runners(sess=sess, coord=coord)
acc_value_batch = sess.run([acc_value])
print(acc_value_batch)
coord.request_stop()
# wait for threads to stop
coord.join(threads=threads)
sess.close()
```

logs:
```
Caused by op 'block1_conv1_bn/keras_learning_phase', defined at:
  File "/home/arkenstone/PycharmProjects/startdt/face_liveness_detect/model/patch_based_cnn/test_image.py", line 66, in <module>
    classes=2)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/_impl/keras/applications/xception.py", line 161, in Xception
    x = BatchNormalization(name='block1_conv1_bn')(x)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/_impl/keras/engine/topology.py", line 252, in __call__
    output = super(Layer, self).__call__(inputs, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/layers/base.py", line 575, in __call__
    outputs = self.call(inputs, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/_impl/keras/layers/normalization.py", line 109, in call
    training = K.learning_phase()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/_impl/keras/backend.py", line 325, in learning_phase
    phase = array_ops.placeholder(dtype='bool', name='keras_learning_phase')
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/array_ops.py", line 1599, in placeholder
    return gen_array_ops._placeholder(dtype=dtype, shape=shape, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 3091, in _placeholder
    "Placeholder", dtype=dtype, shape=shape, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 2956, in create_op
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1470, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'block1_conv1_bn/keras_learning_phase' with dtype bool
	 [[Node: block1_conv1_bn/keras_learning_phase = Placeholder[dtype=DT_BOOL, shape=<unknown>, _device="/job:localhost/replica:0/task:0/device:GPU:0"]()]]
	 [[Node: Cast_2/_23 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_1603_Cast_2", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]
```

**Comment by andydavis1** [comment]
@fchollet Can you comment on this one? Thanks...

**Comment by fchollet** [comment]
Try to put your evaluation code in a different function. The way you've written it, you are reusing the same graph, which could be hiding some underlying issue in your code. Use `K.clear_session()` at the end of each phase, or better, put them inside graph scopes.

`set_learning_phase` certainly does work, so my guess is that you are somehow reusing parts of the training model during evaluation. Using two separate functions (and graphs) should clear that up.

**Comment by CasiaFan** [comment]
@fchollet Thanks for your reply! I try to use evaluation code in a separated script by loading pre-trained model weights, together with adding `K.clear_session()` at the end of session. But the issue still exists. **But again if I use keras instead of tf.keras, it works!** 

BTW, I check the `K.learning_phase()` and `set_learning_phase` does work. So I think this issue should be about graph reusing as you said. Is there something different should be noted between keras and tf about graph usage?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by zizhaozhang** [comment]
This issue also happens to me. I have tried many different ways to resolve this issue. 
Either K.set_learning_phase(0) or K.learning_phase(): 0 gives random results when loading back from a checkpoint @fchollet

**Comment by fchollet** [comment]
I think it's a bug, I'm looking at it.

**Comment by fchollet** [comment]
I believe it's a tricky import issues. Things are working internally, but break when doing

```python
import tensorflow as tf
keras = tf.keras
```

**Comment by fchollet** [comment]
It's actually a user error. The problem is the following:

- You are using `tf.keras`, in particular the tf.keras BN layer (part of Xception model)
- You are setting the learning phase via `K` which is imported from `keras` (not tf.keras)

Therefore setting the learning phase does nothing to your model: it's acting on a different python module altogether...

The solution is this:

```python
from tensorflow import keras
K = keras.backend
```

**Comment by zizhaozhang** [comment]
@fchollet  In my case I do not have this module bug. My environment is tf 1.3. I am sure I use modules all from the same package.
So I tried 1) using tf.contrib.keras and 2) uses keras (2.12). Both cause random results when evaluation. My code was correct in early tensorflow+keras version (probably 8 months ago). But this time it is not working, I  believe there is some bugs of K.learning_phase()

**Comment by fchollet** [comment]
Then please open a new issue (since this one is not related) and provide a code standalone snippet to reproduce the problem.

**Comment by zizhaozhang** [comment]
@fchollet pls checkout this issue. I believe there is a bug. thanks a lot

**Comment by ymcasky** [comment]
Dear @fchollet 
I think this is a bug,too

Here is my example code

```

import tensorflow as tf
import numpy as np


#from tensorflow.python.keras import  backend
from tensorflow import keras
K = keras.backend

from tensorflow.python.keras.layers import  Input

from tensorflow.python.keras.layers import MaxPooling2D
from tensorflow.python.keras.layers import BatchNormalization
from tensorflow.python.keras.layers import Conv2D, Dense, Flatten
from tensorflow.python.keras.layers import Activation
from tensorflow.python.keras.models import Model

print(tf.__version__)
# Load Data
import cifar10

img_size = 32
num_channels = 3
num_classes = 10
cifar10.maybe_download_and_extract()


class_names = cifar10.load_class_names()


images_train, cls_train, labels_train = cifar10.load_training_data()

images_test, cls_test, labels_test = cifar10.load_test_data()

print("Size of:")
print("- Training-set:\t\t{}".format(len(images_train)))
print("- Test-set:\t\t{}".format(len(images_test)))



img_size_cropped = 24
#%%
def pre_process_image(image, training):
    # This function takes a single image as input,
    # and a boolean whether to build the training or testing graph.
    
    if training:
        # For training, add the following to the TensorFlow graph.

        # Randomly crop the input image.
        image = tf.random_crop(image, size=[img_size_cropped, img_size_cropped, num_channels])

        # Randomly flip the image horizontally.
        image = tf.image.random_flip_left_right(image)
        
        # Randomly adjust hue, contrast and saturation.
        image = tf.image.random_hue(image, max_delta=0.05)
        image = tf.image.random_contrast(image, lower=0.3, upper=1.0)
        image = tf.image.random_brightness(image, max_delta=0.2)
        image = tf.image.random_saturation(image, lower=0.0, upper=2.0)

        # Some of these functions may overflow and result in pixel
        # values beyond the [0, 1] range. It is unclear from the
        # documentation of TensorFlow 0.10.0rc0 whether this is
        # intended. A simple solution is to limit the range.

        # Limit the image pixels between [0, 1] in case of overflow.
        image = tf.minimum(image, 1.0)
        image = tf.maximum(image, 0.0)
    else:
        # For training, add the following to the TensorFlow graph.

        # Crop the input image around the centre so it is the same
        # size as images that are randomly cropped during training.
        image = tf.image.resize_image_with_crop_or_pad(image,
                                                       target_height=img_size_cropped,
                                                       target_width=img_size_cropped)

    return image

def pre_process(images, training):
    # Use TensorFlow to loop over all the input images and call
    # the function above which takes a single image as input.
    images = tf.map_fn(lambda image: pre_process_image(image, training), images)

    return images

#%%
def _bn_relu(input_layer):
    """Helper to build a BN -> relu block
    """
    norm = BatchNormalization(axis=3)(input_layer)
    return Activation("relu")(norm)
#%%

x = tf.placeholder(tf.float32, shape=[None, img_size, img_size, num_channels], name='x')

distorted_images = pre_process(images=x, training=True)

y_true = tf.placeholder(tf.float32, shape=[None, num_classes], name='y_true')

y_true_cls = tf.argmax(y_true, axis=1)

def create_network(training):
    # Wrap the neural network in the scope named 'network'.
    # Create new variables during training, and re-use during testing.
    with tf.variable_scope('network', reuse=not training):

        images = x        
        images = pre_process(images=images, training=training)
        
        inputs = Input(tensor=images)

        net_ks = Conv2D(kernel_size=5, strides=1, filters=64, padding='same',
             activation='linear', name='layer_conv1')(inputs)

        net_ks = _bn_relu(net_ks)
        net_ks = MaxPooling2D(pool_size=2, strides=2)(net_ks)


        net_ks = Conv2D(kernel_size=5, strides=1, filters=64, padding='same',
        activation='relu', name='layer_conv2')(net_ks) 
        net_ks = MaxPooling2D(pool_size=2, strides=2)(net_ks)

        
        net_ks = Flatten()(net_ks)
        
        net_ks = Dense(256, activation='relu', name='layer_fc1')(net_ks)
        net_ks = Dense(128, activation='relu', name='layer_fc2')(net_ks)
        preds_ks = Dense(num_classes, activation='linear')(net_ks)
        
        preds_softmax = tf.nn.softmax(preds_ks)
        step1 = tf.cast(y_true, tf.float32) * tf.log(preds_softmax)
        step2 = -tf.reduce_sum(step1, reduction_indices=[1])
        loss = tf.reduce_mean(step2)       # loss
        
        model = Model(inputs=inputs, outputs=preds_ks)

    return preds_softmax, loss, model

#%% Create Neural Network for Training Phase
global_step = tf.Variable(initial_value=0,
                          name='global_step', trainable=False)

y_pred, loss, model = create_network(training=True)

model.summary()

optimizer = tf.train.AdamOptimizer(learning_rate=1e-4).minimize(loss, global_step=global_step)
#%% Create Neural Network for Test Phase / Inference
#y_pred, _ , _= create_network(training=False)
y_pred_cls = tf.argmax(y_pred, axis=1)
correct_prediction = tf.equal(y_pred_cls, y_true_cls)
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

session = tf.Session()
session.run(tf.global_variables_initializer())
train_batch_size = 64

def random_batch():
    # Number of images in the training-set.
    num_images = len(images_train)

    # Create a random index.
    idx = np.random.choice(num_images,
                           size=train_batch_size,
                           replace=False)

    # Use the random index to select random images and labels.
    x_batch = images_train[idx, :, :, :]
    y_batch = labels_train[idx, :]

    return x_batch, y_batch

def optimize(num_iterations):

    for i in range(num_iterations):

        x_batch, y_true_batch = random_batch()


        feed_dict_train = {x: x_batch,
                           y_true: y_true_batch, K.learning_phase(): 1}


        i_global, _ = session.run([global_step, optimizer],
                                  feed_dict=feed_dict_train)

        # Print status to screen every 100 iterations (and last).
        if (i_global % 100 == 0) or (i == num_iterations - 1):
            # Calculate the accuracy on the training-batch.
            feed_dict_train = {x: x_batch,
                           y_true: y_true_batch, K.learning_phase(): 0}
            batch_acc = session.run(accuracy,
                                    feed_dict=feed_dict_train)

            # Print status.
            msg = "Global Step: {0:>6}, Training Batch Accuracy(phase 0): {1:>6.1%}"
            print(msg.format(i_global, batch_acc))
            
            feed_dict_train = {x: x_batch,
                           y_true: y_true_batch, K.learning_phase(): 1}
            batch_acc = session.run(accuracy,
                                    feed_dict=feed_dict_train)

            # Print status.
            msg = "Global Step: {0:>6}, Training Batch Accuracy(phase 1): {1:>6.1%}"
            print(msg.format(i_global, batch_acc))  


#%%
optimize(num_iterations=10000)
```
The output of" Training Batch Accuracy(phase 0) "& "Training Batch Accuracy(phase 1) " is different.
Please help us to use it. Thank you!

**Comment by Red-Eyed** [comment]
Same problem.
I've answered on [SO](https://stackoverflow.com/a/54309538/3049753)
